### PR TITLE
Offline page

### PR DESCRIPTION
**Pull request by bjorvack**
## Type

- Feature

## Pull request description

Adds service worker for offline page
Internal reference https://app.activecollab.com/108877/reports/team-timeline?modal=Task-84793-62

### CODE REVIEW

**Review by Katrienvh — CHANGES_REQUESTED** [review]

**Comment by carakas** [comment]
@bjorvack can you include a screenshot or something so we can see what it looks like?

File: app/Kernel.php [hunk 1]
```diff
@@ -100,6 +100,8 @@ public function defineForkConstants(): void
         defined('ACTION_GROUP_TAG') || define('ACTION_GROUP_TAG', $container->getParameter('action.group_tag'));
         defined('ACTION_RIGHTS_LEVEL') || define('ACTION_RIGHTS_LEVEL', $container->getParameter('action.rights_level'));
 
+        defined('OFFLINE_PATH') || define('OFFLINE_PATH', PATH_WWW . '/offline');
```
Comment (carakas) [inline app/Kernel.php:103]: We are working on getting those constants out, not a fan of adding new ones

File: offline/.gitignore [hunk 2]
```diff
@@ -0,0 +1,2 @@
+*
```
Comment (carakas) [inline offline/.gitignore:1]: why is this here? You have a file in that directory called offline.html but you are also telling git to ignore it?

File: offline/offline.html [hunk 3]
```diff
@@ -0,0 +1,33 @@
+<!doctype html>
```
Comment (carakas) [inline offline/offline.html:1]: why a html page like this? Would it not be possible to include it in the cms? so the text won't need to be hardcoded?

File: offline/offline.html [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+<html lang="en">
+<head>
+  <meta charset="utf-8">
+
+  <title>Offline page</title>
+  <meta name="description" content="Offline page of this awesome website">
+  <meta name="author" content="Sarah Meysman">
```
Comment (carakas) [inline offline/offline.html:9]: I would remove this line

File: offline/offline.html [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  <link href="/src/Frontend/Themes/Fork/Core/Layout/Css/screen.css?m=1593680880" rel="stylesheet" />
+
+</head>
+
+<body>
+<div class="container">
+  <div class="row justify-content-center mt-5">
+    <div class="col-md-8 col-xl-6">
+
+      <h1>You're offline</h1>
```
Comment (carakas) [inline offline/offline.html:21]: wouldnt it be possible to use the translation system so these text are in the language you are visiting the site?

File: package-lock.json [hunk 6]
```diff
@@ -1,8 +1,16451 @@
 {
   "name": "forkcms",
   "version": "1.0.0",
-  "lockfileVersion": 1,
+  "lockfileVersion": 2,
   "requires": true,
+  "packages": {
```
Comment (carakas) [inline package-lock.json:6]: why is there suddenly so much npm stuff here? seems to contain a bunch of things that arent related

File: src/Backend/Modules/Pages/Actions/Settings.php [hunk 7]
```diff
@@ -22,32 +33,95 @@ public function execute(): void
...
+    private function renderOfflinePage(): void
+    {
+        /** @var Environment $twig */
+        $twig = $this->get('twig');
+
+        /** @var ParseBlocksExtension $extension */
+        $extension = $twig->getExtension(ParseBlocksExtension::class);
+
+        $html = $twig->render(
+            BACKEND_MODULES_PATH . '/' . $this->getModule() .'/Layout/Templates/Offline.html.twig',
```
Comment (carakas) [inline src/Backend/Modules/Pages/Actions/Settings.php:101]: wouldn't it make more sense to put this in the frontend? That way you can also override it in your theme

File: src/Backend/Modules/Pages/Actions/Settings.php [hunk 8]
```diff
@@ -22,32 +33,95 @@ public function execute(): void
...
+                'image' => $this->get('media_library.repository.group')->findOneById(
+                    $this->get('fork.settings')->get(
+                        $this->getModule(),
+                        'offline_image_' . Language::getWorkingLanguage()
+                    )
+                )
+            ]
+        );
+
+        $offlinePage = fopen(OFFLINE_PATH . '/offline_' . Language::getWorkingLanguage() . '.html', 'w');
```
Comment (carakas) [inline src/Backend/Modules/Pages/Actions/Settings.php:123]: Please use Filesystem  dump of symfony for this.
Looks like the page is generated but you committed the generated code. Can you have a look at removing it?
Is this also enabled by default? If so if it generated when installing fork?

File: src/Backend/Modules/Pages/Domain/Page/SettingsDataTransferObject.php [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+use Backend\Core\Language\Locale;
+use Backend\Modules\MediaLibrary\Domain\MediaGroup\MediaGroup;
+use Backend\Modules\MediaLibrary\Domain\MediaGroup\MediaGroupRepository;
+use Backend\Modules\MediaLibrary\Domain\MediaGroup\Type as MediaGroupType;
+use Common\ModulesSettings;
+
+class SettingsDataTransferObject
+{
+    const MODULE = 'Pages';
```
Comment (carakas) [inline src/Backend/Modules/Pages/Domain/Page/SettingsDataTransferObject.php:13]: not a big fan of doing this

File: src/Backend/Modules/Pages/Domain/Page/SettingsDataTransferObject.php [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+    public $offlineText;
+
+    /** @var MediaGroup */
+    public $offlineImage;
+
+    public function __construct(ModulesSettings $settings, MediaGroupRepository $mediaGroupRepository)
+    {
+        $this->metaNavigation = $settings->get(
+            $this::MODULE,
```
Comment (carakas) [inline src/Backend/Modules/Pages/Domain/Page/SettingsDataTransferObject.php:27]: why $this and not self if you use this?

File: src/Backend/Modules/Pages/Domain/Page/SettingsDataTransferObject.php (hunk 9 above)
Comment (carakas) [reply src/Backend/Modules/Pages/Domain/Page/SettingsDataTransferObject.php:13]: Please also add public protected or private if you do this

File: src/Backend/Modules/Pages/Layout/Templates/Offline.html.twig [hunk 11]
```diff
@@ -0,0 +1,32 @@
+<!doctype html>
+
+<html lang="en">
+<head>
+  <meta charset="utf-8">
+
+  <title>Offline page</title>
```
Comment (carakas) [inline src/Backend/Modules/Pages/Layout/Templates/Offline.html.twig:7]: Use {{title}} here as well

File: src/Backend/Modules/Pages/Layout/Templates/Offline.html.twig [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+  <meta name="description" content="Offline page of this awesome website">
+  <meta name="author" content="Sarah Meysman">
```
Comment (carakas) [inline src/Backend/Modules/Pages/Layout/Templates/Offline.html.twig:9]: Remove those two lines please

File: src/Backend/Modules/Pages/Layout/Templates/Offline.html.twig [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+
+  <!--<link href="css/bootstrap.min.css" rel="stylesheet" />-->
+  <link href="/src/Frontend/Themes/Fork/Core/Layout/Css/screen.css?m=1593680880" rel="stylesheet" />
```
Comment (carakas) [inline src/Backend/Modules/Pages/Layout/Templates/Offline.html.twig:12]: You should load the css in a proper way, you hardcoded the theme and the timestamp

File: src/Frontend/Core/Header/Header.php [hunk 14]
```diff
@@ -125,6 +126,23 @@ public function __construct(KernelInterface $kernel)
         $this->jsData = new JsData($jsData);
 
         $this->meta = new MetaCollection();
+
+        $this->addJS('/src/Frontend/Modules/Pages/Js/Pages.js', true, true, Priority::module());
```
Comment (carakas) [inline src/Frontend/Core/Header/Header.php:130]: Not a fan of this, include it in the ES6 instead of this way.

**Review by carakas — CHANGES_REQUESTED** [review]

**Comment by carakas** [comment]
@bjorvack could you have a look at the feedback please?

**Comment by carakas** [comment]
Closing stale PR, let me know if you want to finish the feedback

**Pull request closed without merging**